cmd/proxy/console.go

```diff
@@ -0,0 +1,266 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/creack/pty"
+	"github.com/gin-gonic/gin"
+	"github.com/google/uuid"
+	socketio "github.com/googollee/go-socket.io"
+	"net/http"
+	"os"
+	"os/exec"
+	"sync"
+)
+
+var (
+	consoleMap = make(map[string]Console)
+	mutex = sync.Mutex{}
+)
+
+type Console struct {
+	Id      string   `json:"id"`
+	Network string   `json:"network"`
+	Pty     *os.File `json:"-"`
+}
+
+func init() {
+	router.GET("/api/v1/consoles", listConsoles)
+	router.GET("/api/v1/consoles/:id", getConsole)
+	router.POST("/api/v1/consoles", createConsole)
+}
+
+func findById(id string) *Console {
+	console, ok := consoleMap[id]
+	if !ok {
+		return nil
+	}
+	return &console
+}
+
+func getConsole(c *gin.Context) {
+	id := c.Param("id")
+	console := findById(id)
+	if console == nil {
+		c.JSON(http.StatusNotFound, gin.H{})
+		return
+	}
+	c.JSON(http.StatusOK, console)
+}
+
+func listConsoles(c *gin.Context) {
+	c.JSON(http.StatusOK, consoleMap)
+}
+
+type TerminalSize struct {
+	Rows int `json:"rows"`
+	Cols int `json:"cols"`
+}
+
+type StartRequest struct {
+	Id   string       `json:"id"`
+	Size TerminalSize `json:"size"`
+}
+
+type StopRequest struct {
+	Id string `json:"id"`
+}
+
+func createConsole(c *gin.Context) {
+	network := os.Getenv("NETWORK")
+	id := fmt.Sprint(uuid.New())
+	console := Console{
+		Id:      id,
+		Network: network,
+	}
+	consoleMap[id] = console
+	e := fmt.Sprintf("console-%s", id)
+	logger.Debugf("[console] Register event %s", e)
+	c.JSON(http.StatusOK, console)
+}
+
+func writeInitScript(network string) {
+	f, err := os.Create("init.bash")
+	if err != nil {
+		logger.Errorf("Failed to write init.bash: %s", err)
+		return
+	}
+	defer f.Close()
+
+	f.WriteString(`\
+cat <<EOF
+
+                           .___           __  .__   
+          ___  _____ __  __| _/     _____/  |_|  |  
+          \  \/  /  |  \/ __ |    _/ ___\   __\  |  
+           >    <|  |  / /_/ |    \  \___|  | |  |__
+          /__/\_ \____/\____ |     \___  >__| |____/
+                \/          \/         \/           
+--------------------------------------------------------------
+
+EOF
+
+export NETWORK=` + network + `
+export PS1="$NETWORK > "
+function start() {
+	docker start ${NETWORK}_${1}_1 
```

Comment:
Is the suffix always going to be "_1"?

Reply:
@erkarl It will be 1 in our use case for a long time. And we definitely should migrate to docker-compose instead of docker.